### PR TITLE
Add support for building for {i686,x86_64}-pc-windows-msvc using clang/clang-cl 12.

### PR DESCRIPTION
**Pull request by briansmith**
See each individual commit message for details. We don't have CI/CD coverage for this build configuration yet, but I tested it locally.

### CODE REVIEW

**Comment by briansmith** [comment]
@awakecoding @Alovchin91 Please review this. This is one of many changes I'm making to officially support Clang for -msvc targets. The build.rs stuff is one of two changes that should make it possible to eliminate the special `clang()` function in your PR, and it will also allow the user to use a `clang` executable that isn't named `clang`, e.g. `clang-12` on some of my systems.

**Review by awakecoding — APPROVED** [review]
those changes look good to me

**Review by Alovchin91 — APPROVED** [review]
I generally love everything that removes extra code and uses suitable types 😛

File: build.rs [hunk 1]
```diff
@@ -508,14 +503,28 @@ fn build_library(
...
+    out_dir: &Path,
+) -> String {
     let ext = p.extension().unwrap().to_str().unwrap();
     if ext == "o" {
         p.to_str().expect("Invalid path").into()
     } else {
         let out_path = obj_path(out_dir, p);
         let cmd = if target.os != WINDOWS || ext != "asm" {
-            cc(p, ext, target, warnings_are_errors, &out_path, out_dir)
+            cc(
```
Comment (Alovchin91) [inline build.rs:519]: There's still a call in `pregenerate_asm_main` that calls `nasm` directly though

**Review by Alovchin91 — CHANGES_REQUESTED** [review]

File: build.rs (hunk 1 above)
Comment (briansmith) [reply build.rs:519]: > There's still a call in pregenerate_asm_main that calls nasm unconditionally though.

I'm making more changes to address that in upcoming PRs.

**Review by Alovchin91 — APPROVED** [review]

**Pull request merged**